### ISSUE TITLE
Building a Spreadsheet: Logic for avoiding self-reference in function calls is broken

### ISSUE THREAD

**Issue by lasjorg** [freeCodeCamp/freeCodeCamp#55060]
### Describe the Issue

In step 31 we are asked to add a check that the value doesn't contain the id. As far as I can tell, this is to avoid a function call containing its own cell value.

However, the ids are uppercase, and the check is not converting the value.

```
A1 = 10
A2 = 10
A3 = =sum(a1, a2) -> 20
```

```
A1 = 10
A2 = 10
A3 = =sum(a1, a3) -> undefined (and Potential infinite loop in console)
```

But if you `toUpperCase` the value, that doesn’t happen.

Edit: As also pointed out in the thread, `includes` will match on partial strings. Causing a false positive with the value converted.

`!"A11".includes("A1")`

```
A1 = =sum(a2, a11)
A2 = 10
A11 = 10
```

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-functional-programming-by-building-a-spreadsheet/step-31

### Your code

```



```

### Expected behavior

Calling one of the functions with its own cell reference passed in as an argument should not run.

### System

- Device: [e.g. iPhone 6, Laptop]
- OS: [e.g. iOS 14, Windows 10, Ubuntu 20.04]
- Browser: [e.g. Chrome, Safari]
- Version: [e.g. 22]

### Additional context

https://forum.freecodecamp.org/t/learn-functional-programming-by-building-a-spreadsheet-step-31/693224

**Comment by naomi-lgbt** [comment]
Yes, this should be fixed. We could definitely update the text to something like:

> Now you need to check if the `value` does not include the `id` of the element. Create an `if` condition to do so. Your condition will need to convert the `value` to upper case characters.

I'm not entirely sure how we want to resolve the false positives.

**Comment by lasjorg** [comment]
> I'm not entirely sure how we want to resolve the false positives.

Not sure if there is an easy solution to that. The code also doesn't guard against passing in refrences to empty cells or calling with no arguments.

```
=sum(a1, a2) -> NaN

=sum() -> NaN
```

I'm not even sure what all the valid and invalid inputs are.

Without having deciphered all the function calls and functions returning functions (which is slightly headache inducing). It seems to me like the `applyFunction` or `evalFormula` functions should be responsible for guarding their input, and not the caller at the callsite.

**Comment by jdwilkin4** [comment]
closing since this is archived coursework
